telemetry/firmware/telemetry.c

```diff
@@ -0,0 +1,316 @@
+#define BOTS_FAIL                       2
+#define COCKPIT_ESTOP_FAIL              3
+#define HV_DISCONNECT_FAIL              4
+#define HVD_CONN_FAIL                   5
+#define PACK_CONN_FAIL                  6
+#define BMS_FAIL                        7
+#define IMD_FAIL                        8
+#define TSMS_FAIL                       9
+
+
+/*----- Global Variables -----*/
+
+volatile uint8_t gFlag = 0x00; // global flag
+
+volatile uint16_t shutdown_error_flag = 0x00; // system failure error code flag - default 0x00. 
+
+// CAN DATA to store & send out over UART 
+uint8_t accumulator_voltages[192] = {0}; // 192 wholeass bytes wtf
+uint8_t accumulator_temperatures[80] = {0}; 
+uint16_t throttle = 0x00;
+uint8_t state_of_charge = 0x00; 
+
+
+uint8_t can_recv_msg[8] = {0}; // CAN msg recieved
+
+
+
+uint8_t data_byte = 0x00; // for UART 
+
+
+
+typedef struct {
+    uint8_t ID; 
+    short data_length; 
+    uint8_t data[8]; 
+} CAN_MSG; 
+
+// QUEUES
+
+QUEUE(UART, uint8_t, UART_QUEUE_SIZE); 
+QUEUE(CAN, CAN_MSG *, CAN_QUEUE_SIZE); 
+
+voltatile struct queue_CAN CAN_queue; 
+voltatile struct queue_UART UART_queue; 
+
+
+
+// Define CAN struct --> to be created and set with ID & data 
+
+
+/*----- Interrupt(s) -----*/
+
+
+ISR(CAN_INT_vect){ // TODO: Write state machine for mailbox address switching
+/*----- Brake light mailbox -----*/
+    CANPAGE = (BRAKE_LIGHT_MBOX<<MOBNB0); // what does this do? 
+    if (bit_is_set(CANSTMOB, RXOK)) {
+		can_recv_msg[0] = CANMSG;   // PANIC PANIC
+		can_recv_msg[1] = CANMSG;   // brake analog voltage MSB       // subjectlines of the letter, have to read the letters in order
+		can_recv_msg[2] = CANMSG;   // brake analog voltage LSB
+		can_recv_msg[3] = CANMSG;   // is brake pressed? CF
+		can_recv_msg[4] = CANMSG;   // BSPD sense
+
+        // HUH ? wtf? below
+
+		can_recv_msg[5] = CANMSG;   // TSMS sense
+		can_recv_msg[6] = CANMSG;   // left e-stop sense
+		can_recv_msg[7] = CANMSG;   // GLVMS sense
+
+        /* For ALL CAN DATA, GENERATE CAN_MSG STRUCT AND PASS INTO CAN_QUEUE*/ 
+
+        CAN_MSG * msg = malloc(sizeof(CAN_MSG));  
+        (*msg) = { custom_ID, 8, can_recv_msg}; // TODO: To confirm, this copies over can_recv_mesg during asisgnment?
+
+        if (queue_CAN_push(&CAN_queue, msg) == 0) {
+            //push successful, no action required
+        } else {
+            //not enough room in queue, handle appropriately here
+        }
+        
+        // TODO: what do we do about sending the shutdown error flag? It's not a CAN struct? Should we check for it on every send and if it's ! 0 we send it? 
+
+
+
+		if(can_recv_msg[4] == 0xFF) { //if the specific letter says something
+			shutdown_error_flag |= _BV(BSPD_FAIL);           //trip flag // scribbling on your hand
+		} else {
+			shutdown_error_flag &= ~_BV(BSPD_FAIL);
+		}
+		//Setup to Receive Again for future
+		CANSTMOB = 0x00;
+		CAN_wait_on_receive(BRAKE_LIGHT_MBOX, CAN_ID_BRAKE_LIGHT, CAN_LEN_BRAKE_LIGHT, CAN_MSK_SINGLE);
+	}
```

Comment:
I know I said not to use malloc above, but just for the sake of completeness, you *must* free dynamically allocated memory at the end of usage. Call `free(msg)` (but actually just don't use malloc).